### PR TITLE
Alejandro/squarelotron

### PR DESCRIPTION
**Pull request by Sl3v7n**

### CODE REVIEW

**Comment by carledriss** [comment]
@Sl3v7n 
Please update your branch against develop

File: src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        assert Arrays.deepEquals(actual, expected);
+    }
+
+    /**
+     * This method will test a diagonal flip with size 5 on the second ring.
+     */
+    @Test
+    public void testMainDiagonalFlipSize5With2Ring() {
+
+        final int[][] expected = {{1, 2, 3, 4, 5}, {6, 7, 12, 17, 22}, {11, 8, 13, 18, 15},
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java]: this expected value is wrong
it should be:
```java
        final int[][] expectedMatrix = {{1, 2, 3, 4, 5}, {6, 7, 12, 17, 10}, {11, 8, 13, 18, 15}, {16, 9, 14, 19, 20}, {21, 22, 23, 24, 25}};
```

File: src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+     */
+    @Test
+    public void testMainDiagonalFlipDimensionFiveRingThree() {
+        //null
+
+        final int localSize = 5;
+        final int ring = 3;
+        Squarelotron square = new Squarelotron(localSize);
+        actualSquarelotron = square.mainDiagonalFlip(ring);
+        final int[][] expected = {{1, 2, 3, 4, 5}, {6, 7, 8, 9, 10}, {11, 12, 13, 18, 23},
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java]: this expected value is wrong
it should be:
```java
        final int[][] expectedMatrix = {{1, 2, 3, 4, 5}, {6, 7, 8, 9, 10}, {11, 12, 13, 14, 15}, {16, 17, 18, 19, 20}, {21, 22, 23, 24, 25}};
```

**Comment by carledriss** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/main/java/org/minions/devfund/alejandro/Squarelotron.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
+package org.minions.devfund.alejandro;
+
+/**
+ * Squareotron class.
+ */
+public class Squarelotron {
+
+    private int[][] squarelotron;
```
[the remark this answers is not in the document]
Comment (carledriss) [reply src/main/java/org/minions/devfund/alejandro/Squarelotron.java]: For more information review
https://sonarcloud.io/coding_rules?open=squid%3AS1700&rule_key=squid%3AS1700

**Comment by carledriss** [comment]
@Sl3v7n 
Don't forget to update your branch against develop
Also don't forget to add screenshots about your progress

File: src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+
+
+    /**
+     * This method will initialize the Squarelotron class.
+     */
+    @Before
+    public void setUp() {
+        final int size = 5;
+        squarelotron =  new Squarelotron(size);
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java]: The setup method is created when it is reused the attributes of class in all the tests
So in this case you are not reusing squarelotron instance with size 5 in all your test.
Please fix this problem.

File: src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+import java.util.Arrays;
+
+/**
+ * Class that will test the class Squarelotron.
+ */
+public class SquarelotronTest {
+
+    private Squarelotron squarelotron;
+    private Squarelotron actualSquarelotron;
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java]: Is it really needed to have those attributes of class ?

**Comment by carledriss** [comment]
@Sl3v7n 
Also don't forget to increase the code coverage
https://codecov.io/gh/Minions-F/devfund/pull/21/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9vcmcvbWluaW9ucy9kZXZmdW5kL2FsZWphbmRyby9TcXVhcmVsb3Ryb24uamF2YQ==

It is missing the scenarios when the ring is not greater than zero

**Comment by Sl3v7n** [comment]
image
image

File: src/test/java/org/minions/devfund/alejandro/tests/WhackAMoleTest.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+import org.minions.devfund.alejandro.WhackAMole;
+import org.junit.Test;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * This class contain the test cases for WhackAMole class.
+ */
+
+public class WhackAMoleTest {
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/WhackAMoleTest.java]: This class should be deleted in this pull request

File: src/main/java/org/minions/devfund/alejandro/Squarelotron.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/Squarelotron.java]: MAJOR Rename field "squarelotron" [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1700)

File: src/main/java/org/minions/devfund/alejandro/Squarelotron.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+     * @return A modified instance of a new Squarelotron.
+     */
+    public Squarelotron upsideDownFlip(int ring) {
+
+        Squarelotron resultSquare = new Squarelotron(this.size);
+        int[][] swappedMatrix = this.getSquarelotron();
+
+        if (ring > 0 && ring <= this.size / 2) {
+
+            swappedMatrix = swapMatrix(swappedMatrix);
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/Squarelotron.java]: MAJOR Remove this useless assignment; "swappedMatrix" already holds the assigned value along all execution paths. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)

**Comment by carledriss** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
+package org.minions.devfund.alejandro.tests;
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/SquarelotronTest.java]: In fact your unit tests should be in the package 
package org.minions.devfund.alejandro

tests package is not necessary

File: src/main/java/org/minions/devfund/alejandro/Squarelotron.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+     * @param ring the ring that will be flipped in the square.
+     * @param squarelotronFlipedMatrix is the matrix that will be flipped.
+     * @param swapedMatrix the swapped matrix.
+     * @return the flipped matrix.
+     */
+    private int[][] flipSquarelotron(int ring, final int[][] squarelotronFlipedMatrix, final int[][] swapedMatrix) {
+
+        for (int i = ring - 1; i <= this.size - ring; i++) {
+            for (int j = ring - 1; j <= this.size - ring; j++) {
+                if (i == ring - 1 || j == ring - 1 || i == this.size - ring || j == this.size - ring) {
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/Squarelotron.java]: This if condition is repeated in line 99
Can you add a private method called isPositionInRing() ?

File: src/main/java/org/minions/devfund/alejandro/Squarelotron.java [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+        int[][] swappedMatrix = this.getSquarelotronMatrix();
+
+        if (ring > 0 && ring <= this.size / 2) {
+
+            swapMatrix(swappedMatrix);
+            resultSquare.squarelotronMatrix =
+                    flipSquarelotron(ring, resultSquare.getSquarelotronMatrix(), swappedMatrix);
+        }
+        return resultSquare;
+    }
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/Squarelotron.java]: This code may be refactored 
something like this,
```java
    /**
     * This method performs the Upside-Down Flip of the squarelotronMatrix,
     * and returns the new squarelotronMatrix. The original squarelotronMatrix should not be modified.
     * @param ring the ring that will be flipped.
     * @return A modified instance of a new Squarelotron.
     */
    public Squarelotron upsideDownFlip(int ring) {
        Squarelotron resultSquare = new Squarelotron(this.size);
        if (ring > 0 && ring <= this.size / 2) {
            resultSquare.squarelotronMatrix = flipSquarelotron(ring, resultSquare.getSquarelotronMatrix());
        }
        return resultSquare;
    }

    /**
     * This method will flip a matrix in a determined ring.
     * @param ring the ring that will be flipped in the square.
     * @param squarelotronMatrix is the matrix that will be flipped.
     * @return the flipped matrix.
     */
    private int[][] flipSquarelotron(int ring, final int[][] squarelotronMatrix) {
        int[][] swappedMatrix = swapMatrix(this.squarelotronMatrix);
        for (int i = ring - 1; i <= this.size - ring; i++) {
            for (int j = ring - 1; j <= this.size - ring; j++) {
                if (isPositionInRing(ring, i, j)) {
                    squarelotronMatrix[i][j] = swappedMatrix[i][j];
                }
            }
        }
        return squarelotronMatrix;
    }

    /**
     * This method will swap a matrix.
     * @param matrix The matrix that will be swapped.
     * @return the same matrix but swapped.
     */
    private int[][] swapMatrix(final int[][] matrix) {
        int controlSize = this.size - 1;
        for (int i = 0; i < this.size / 2; i++) {
            int[] rowUpside = matrix[i];
            int[] rowDown = matrix[controlSize];
            matrix[i] = rowDown;
            matrix[controlSize] = rowUpside;
            controlSize--;
        }
        return matrix;
    }

    private boolean isPositionInRing(int ring, int i, int j) {
        return i == ring - 1 || j == ring - 1 || i == this.size - ring || j == this.size - ring;
    }
```

**Comment by carledriss** [comment]
Good job @Sl3v7n 
👍

**Comment by carledriss** [comment]
LGTM! 👍

**Pull request merged**